platooning_strategic_IHP/src/platoon_strategic_ihp.cpp

```diff
@@ -1157,207 +906,1233 @@ namespace platoon_strategic
+        bool isCurrentStateTimeout = (tsStart - candidatestateStartTime) > waitingStateTimeout * 1000;
+        ROS_DEBUG_STREAM("timeout1: " << tsStart - candidatestateStartTime);
+        ROS_DEBUG_STREAM("waitingStateTimeout: " << waitingStateTimeout * 1000);
+        if (isCurrentStateTimeout) {
+            ROS_DEBUG_STREAM("The current leader aborting state is timeout. Change back to leader state.");
+            pm_.current_platoon_state = PlatoonState::LEADER;
+        }
+
+        // Task 2: plan timeout
+        if (pm_.current_plan.valid) {
+            {
+                if (pm_.current_plan.valid) {
+                    ROS_DEBUG_STREAM("pm_.current_plan.planStartTime: " << pm_.current_plan.planStartTime);
+                    ROS_DEBUG_STREAM("timeout2: " << tsStart - pm_.current_plan.planStartTime);
+                    ROS_DEBUG_STREAM("NEGOTIATION_TIMEOUT: " << NEGOTIATION_TIMEOUT);
+                    bool isPlanTimeout = (tsStart - pm_.current_plan.planStartTime) > NEGOTIATION_TIMEOUT;
+                    if (isPlanTimeout) {
+                        pm_.current_plan.valid = false;
+                        ROS_DEBUG_STREAM("The current plan did not receive any response. Abort and change to leader state.");
+                        pm_.current_platoon_state = PlatoonState::LEADER;
+                        ROS_DEBUG_STREAM("Changed the state back to Leader");
+                    }
+                }
+            }
+        }
+
+        // Task 3: update plan: PLATOON_FRONT_JOIN with new gap
+        double desiredJoinGap2 = config_.desiredJoinTimeGap * current_speed_;
+        double maxJoinGap = std::max(config_.desiredJoinGap, desiredJoinGap2);
+        // check if compatible for front join --> return front gap, no veh type check, is compatible
+        double currentGap = pm_.getDistanceToFrontVehicle();
+        ROS_DEBUG_STREAM("Based on desired join time gap, the desired join distance gap is " << desiredJoinGap2 << " ms");
+        ROS_DEBUG_STREAM("Since we have max allowed gap as " << desiredJoinGap << " m then max join gap became " << maxJoinGap << " m");
+        ROS_DEBUG_STREAM("The current gap from radar is " << currentGap << " m");
+        // TODO: temporary
+        if (true)//(currentGap <= maxJoinGap && pm_.current_plan.valid == false) {
+        {
+            // compose frontal joining plan, senderID is the old leader 
+            cav_msgs::MobilityRequest request;
+            std::string planId = boost::uuids::to_string(boost::uuids::random_generator()());
+            long currentTime = ros::Time::now().toNSec() / 1000000;
+            request.header.plan_id = planId;
+            request.header.recipient_id = pm_.targetLeaderId;
+            request.header.sender_bsm_id = host_bsm_id_;
+            request.header.sender_id = config_.vehicleID;
+            request.header.timestamp = currentTime;
+
+            // change to new plan type --> add corresponding plantype to cav_msgs/PlanType.msg
```

Comment:
We will need another PR in the Msgs repo to cover this and any other message structure changes before this will run.  I have added one or two other comments in this PR that may affect thinking on msg structure.